### PR TITLE
Disallow multiple form submissions

### LINKED ISSUE #28: Disallow multiple form submissions
Store a `has_applied` Boolean or `applied_at` Date in Mongo to check whether user has already submitted an application. 

If User has already applied, the solution can be **one** of the following:
- Show an error if they have already applied
- Display a message on the form ("Your application has been submitted" ?) and disable form
- Show an alternate view (date they submitted application and/or other details?). An alternate view may allow us to support other features.

### PR DESCRIPTION
**Pull request by jomasti**
This repo doesn't seem to be updated much anymore, but I thought I would give this a shot anyway. I at least got reacquainted with `express`.

Closes #28

### CODE REVIEW

File: routes/applied/index.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import assign from 'lodash/assign'
+import get from 'lodash/get'
+import validate from '../../lib/validate'
+import { getStrings } from '../../lib/helpers'
+import User from '../../lib/db/models/user'
+import express from 'express'
+import bug from 'debug'
+
+const debug = bug('SIR:thanks')
```
Comment (therealklanni) [inline routes/applied/index.js]: `SIR:applied`?

File: routes/apply/index.js [hunk 2]
```diff
@@ -30,6 +30,10 @@ router.get('/', validate, (req, res) => {
       return res.redirect('/')
     }
 
+    if (user.applied_at) {
```
Comment (therealklanni) [inline routes/apply/index.js:39]: What do you suggest for existing users who have already applied?

**Review by therealklanni — CHANGES_REQUESTED** [review]
Looks good, however I suggest we come up with a solution for existing users. Or possibly present a slightly different message when they land on /apply and do not have `applied_at`. We may need to check the `created_at` date and make an educated guess that before a certain date (or time-frame, e.g. "1 month ago") they likely have applied. The messaging could be something along the lines of "Hmm, we are unable to determine whether you have previously applied, but we are reasonably certain you have."

There's also the issue of what should a person who has already applied do in the situation that they

- have already applied, but were not yet accepted (maybe if `applied_at` is <48hrs, we politely ask them to wait a bit longer, or if it's >48hrs we resend the Slack post and notify them the admins have been reminded of their request? In the latter case we may want to limit this to a max number of reminder reposts, and possibly even track when the last reminder was posted)
- have already applied, but will not be receiving an invite (this will require a way for admins to update users with information like an `invited_at` date or `denied_at`, or something along these lines)

Should we present a solution for those cases for the user to contact us or post a new message to Slack such as "Person X has already applied but requires assistance" as a way to inform the admins to reach out to them? I'm not sure what the best UX here would be.

Feel free to contact me on AZWD Slack if you'd like to discuss further! If we can't come up with at least an easy temporary solution for existing users that have already applied, we can merge this as-is anyway, as it's at least better than nothing. Thanks!

**Comment by jomasti** [comment]
I think your suggestions make sense. The one issue I see with checking `created_at` is that that field looks to be the date that the Github profile was created rather than the date the user record was created. In the authorization code, it looks to attempt to set the `createdAt` field when updating the document, but that field isn't in the model, so it's not saved. Should that be changed to use the right field name in order to continue on with that? Seems like it would still be an issue for existing users.

As for the users that have previously applied, I like the idea of a gentle reminder if it's still within 48 hours. It would make more sense if there was an `invited_at` field already there to be confident in display that message. Perhaps there needs to be a PR to add both `invited_at` and `denied_at` fields before continuing with this? Another option is to create a new issue for this particular feature. Or I could still do it all in the same PR, if that's preferred.

File: routes/applied/index.js (hunk 1 above)
Comment (jomasti) [reply routes/applied/index.js]: Whoops!

File: routes/apply/index.js (hunk 2 above)
Comment (jomasti) [reply routes/apply/index.js:39]: I liked your suggestion of assuming a prior application based on a certain timeframe. There's just the issue I raised in my main comment.

**Comment by therealklanni** [comment]
You're correct, I'd forgotten that. `updated_at` is what I meant, this is the one that tracks when a record of you was last changed in our system. If you sign in with GitHub and never come back after applying, that date will be in most cases the same as your apply date. `updated_at` should only ever change if you sign in again with GitHub (your session expired and you came back later) and if I remember correctly, only if any of your GH info has changed. `created_at` and `updated_at` are both saved, despite not being in the model. The model just describes the GitHub information we store (so `created_at` should have been included and we can fix that). 

Obviously, if the user signs in again and `updated_at` is updated we can't check that to see when they applied, so that's where `applied_at` will be useful. The problem is just updating users retroactively. I think checking `updated_at` is our best option, even though it's not 100% reliable. Which is why I was also thinking a fallback would be good to have, like "if you think this is in error, click reapply below". We can probably leave that problem for another day, though; your feedback is welcome.

Same PR is fine with me, it's all closely related. We don't need a `denied_at` just yet, though. We'd need some sort of admin panel added before we can work on anything of that nature. However, we've been working on a complete rewrite of the site (https://github.com/az-webdevs/azwebdevs.org), so we probably shouldn't make major changes like that here.

**Pull request closed without merging**